### PR TITLE
Make hook status printing configurable with delay

### LINKED ISSUE #9610: Disable remote messages on push
With the pre- and post-receive calls introduced in #8602, Gitea now sends messages to the pushing client, for example:

```
remote: * Checking 1 branches
remote: Checked 1 references in total
remote: . Processing 1 references
remote: Processed 1 references in total
```

Since git always displays this messages, even when called with `--quiet`, it would be great if these messages could be turned off.

### PR DESCRIPTION
**Pull request by zeripath**
Add 2 configuration options to control printing of push hook status with a configurable delay.

```ini
[git]
...
VERBOSE_PUSH=true
VERBOSE_PUSH_DELAY=5s
```

Fix #9610

### CODE REVIEW

**Review by guillep2k — COMMENTED** [review]
I've written a bunch of comments and suggestions, but I'm not sure this is the right approach. I realize you want to be faithful to the previous state, where a dot was printed for each commit that has been checked. But maybe it would be much easier if we'd just print a dot for each second that has passed after the initial delay (with no goroutines or string builders).

File: cmd/hook.go [hunk 1]
```diff
@@ -139,20 +169,22 @@ Gitea or set your environment appropriately.`, "")
 				case http.StatusOK:
 					// no-op
 				case http.StatusInternalServerError:
+					close(commsChan)
+					<-tickerChan
```
[the remark this answers is not in the document]
Comment (zeripath) [reply cmd/hook.go]: So the <- tickerChan is there to force the rest of the printing if we had printed something.

By closing the commsChan you break the loop in the ticker goroutine. You then need to wait to allow the ticker goroutine to finish it's job before you return from the function. If you don't there is no guarantee that the output that should be printed will be printed.

[the remark this answers is not in the document]
Comment (guillep2k) [reply cmd/hook.go]: I may be wrong, but if we close commsChan, the loop breaks in a controlled way (ok = false), only after the channel was depleted. Instead, if the loop breaks at tickerChan (sounds like an anime character!) some data may remain in commsChan unprocessed.

Note: my suggestion above is dependent on my other suggestions.

**Comment by zeripath** [comment]
I've moved to a wrapped writer implementation with a configurable delay.

**Review by lunny — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Review by guillep2k — APPROVED** [review]
Nice!

File: cmd/hook.go [hunk 2]
```diff
@@ -58,6 +60,81 @@ var (
...
+	stopped := d.timer.Stop()
+	if stopped {
+		return nil
+	}
+	<-d.timer.C
+	if d.buf == nil {
+		return nil
+	}
+	_, err := d.internal.Write(d.buf.Bytes())
+	return err
```
Comment (zeripath) [inline cmd/hook.go:129]:
```suggestion
	d.buf = nil
	return err
```

File: cmd/hook.go [hunk 3]
```diff
@@ -309,7 +402,7 @@ Gitea or set your environment appropriately.`, "")
 			fail("Internal Server Error", "SetDefaultBranch failed with Error: %v", err)
 		}
 	}
-
+	dWriter.Close()
```
Comment (zeripath) [inline cmd/hook.go]:
```suggestion
	_ = dWriter.Close()
```

File: cmd/hook.go [hunk 4]
```diff
@@ -288,19 +382,18 @@ Gitea or set your environment appropriately.`, "")
 	hookOptions.NewCommitIDs = newCommitIDs[:count]
 	hookOptions.RefFullNames = refFullNames[:count]
 
-	fmt.Fprintf(os.Stdout, " Processing %d references\n", count)
-	os.Stdout.Sync()
+	fmt.Fprintf(out, " Processing %d references\n", count)
 
 	resp, err := private.HookPostReceive(repoUser, repoName, hookOptions)
 	if resp == nil {
+		dWriter.Close()
```
Comment (zeripath) [inline cmd/hook.go]:
```suggestion
		_ = dWriter.Close()
```

File: cmd/hook.go [hunk 5]
```diff
@@ -277,9 +371,9 @@ Gitea or set your environment appropriately.`, "")
 				fail("Internal Server Error", "SetDefaultBranch failed with Error: %v", err)
 			}
 		}
-		fmt.Fprintf(os.Stdout, "Processed %d references in total\n", total)
-		os.Stdout.Sync()
+		fmt.Fprintf(out, "Processed %d references in total\n", total)
 
+		dWriter.Close()
```
Comment (zeripath) [inline cmd/hook.go]:
```suggestion
		_ = dWriter.Close()
```

File: cmd/hook.go [hunk 6]
```diff
@@ -250,16 +345,15 @@ Gitea or set your environment appropriately.`, "")
...
 		if count >= hookBatchSize {
-			fmt.Fprintf(os.Stdout, " Processing %d references\n", count)
-			os.Stdout.Sync()
+			fmt.Fprintf(out, " Processing %d references\n", count)
 			hookOptions.OldCommitIDs = oldCommitIDs
 			hookOptions.NewCommitIDs = newCommitIDs
 			hookOptions.RefFullNames = refFullNames
 			resp, err := private.HookPostReceive(repoUser, repoName, hookOptions)
 			if resp == nil {
+				dWriter.Close()
```
Comment (zeripath) [inline cmd/hook.go]:
```suggestion
				_ = dWriter.Close()
```

**Comment by sapk** [comment]
I stopped and restarted the build which seems blocked.

File: cmd/hook.go [hunk 7]
```diff
@@ -58,6 +60,82 @@ var (
...
+
+func (d *delayWriter) Close() error {
+	if d == nil {
+		return nil
+	}
+	stopped := d.timer.Stop()
+	if stopped {
+		return nil
+	}
+	<-d.timer.C
```
Comment (zeripath) [inline cmd/hook.go]: Stupid golang doc implies that the channel is closed on stop - it is not!

```suggestion
	select {
	case <-d.timer.C:
	default:
	}
```

**Comment by zeripath** [comment]
@sapk yup that was my fault. The documentation and example for New timer is particularly bad. I even knew about this problem but forgot about it.

**Comment by zeripath** [comment]
Make lg-tm work

**Pull request merged**